CSV input with metadata support

ivarref opened the pull request:
This pull request is a WIP attempt at adding CSV input support and provide metadata separately in the form of a JSON file.

The intention is that a user should be able to generate SAS/SAV/DTA files from CSV and JSON. An example CSV file would look like this:

```
gender
1
2
9
```

while the metadata would look like this:

```
{
    "type": "STATA",
    "variables": [
        {
            "name": "gender",
            "type": "NUMERIC",
            "label": "gender of respondent",
            "categories": [
                {
                    "code": 1,
                    "label": "Male"
                },
                {
                    "code": 2,
                    "label": "Female"
                },
                {
                    "code": 9,
                    "label": "No answer"
                }
            ],
            "missing": {
                "type": "DISCRETE",
                "values": [
                    9
                ]
            }
        }
    ]
}
```

The metadata has a schema written in [JSON Schema](http://json-schema.org/) so it should be possible to validate the metadata in many languages.

This change requires the [libcsv](https://github.com/robertpostill/libcsv/) library for parsing CSV.

I believe there is some small memory leaks in the existing code. I've tried to fix the simplest of them, and I've also added the make target `check-valgrind` for showcasing the remaining.

Comments or suggestions?

Everything is not complete in the current code, most notably `storage_width` is not implemented yet. It's coming soon.

Thanks.

evanmiller commented:
This is a good start! I see the build is failing though -- I think what you'll need to do is make this compile conditional on the presence of libcsv, similar to how XLSX support is conditional on the presence of libxlsxwriter. Check out Makefile.am around line 99 and configure.ac around line 26. autotools are a world of pain but if you imitate the XLSX support it shouldn't be too hard.

A few more general comments:

I'd like to avoid Python if at all possible, even if it's just for tests.

I'd like to review the Docker / valgrind stuff separately, possibly as a "prerequisite" to the CSV stuff.

I'm not sure the DTA date calculations have the right home, do you envision this being part of the library API or just as a helper for the `readstat` binary?

Some documentation describing the JSON schema is probably in order.

ivarref commented:
Thanks for the feedback Evan.

libxlsxwriter-style: Yes, I will try to make it like that.

Avoid Python: Sure, that should not be a problem.

Docker / valgrind: That's also fine, I just wanted to be sure I could reproduce the error in a consistent manner.

_I'm not sure the DTA date calculations have the right home, do you envision this being part of the library API or just as a helper for the `readstat` binary?_
I didn't think too hard about that I guess. Perhaps a helper function is more appropriate.

I'll look into describing the schema as well.

I'll continue working on Monday. Have a good weekend!

ivarref commented:
It's building now (libxlsxwriter-style). I've removed Docker/valgrind for now.

I've also added a simple metadata extractor --- currently hardcoded for SAV files.

Still remaining: Remove Python (validate schema instance in C?) + various clean ups.

evanmiller commented:
Also please make sure the new tests are executed on both Travis and Appveyor. This can be accomplished by modifying `appveyor.yml` and `.travis.yml`. I will have more specific comments when you're happy with the PR and all the tests are passing.

evanmiller commented:
@ivarref Since this PR is noisy I am muting for now -- please @ mention me when you'd like my attention, thanks!

ivarref commented:
Sure @evanmiller ... Many things are starting to fit into place, but this is also becoming a pretty big change.

ivarref commented:
Hi @evanmiller 
I've added metadata description now --- and I'm happy with the functionality that is added. We have been using this fork of ReadStat in-house at [Norwegian Centre for Research Data](http://www.nsd.uib.no/nsd/english/index.html) for close to a month now with no problems.

Do you have some further feedback? I do know I've taken some shortcuts, e.g. some `exit(1)`s. How far from acceptable do you find the current code?

Thanks.

evanmiller commented:
Thanks @ivarref. Overall it looks good but I will leave specific comments in the code.

One last bit of housekeeping is copyright; is your patch owned by you or Norwegian Centre for Research Data? Should probably make an AUTHORS file for future reference.

evanmiller commented on Makefile.am:
```diff
@@ -70,10 +72,30 @@ noinst_HEADERS = \
+	   src/bin/extract_metadata.h \
+	   src/bin/extract_label_set.h \
+	   src/bin/write_missing_values.h \
+	   src/bin/write_value_labels.h \
```
These should be consistent with tabs/spaces.

evanmiller commented on src/bin/modules/produce_csv_value_dta.c:
```diff
@@ -0,0 +1,106 @@
+                .tag = 'a' + i,
+                .v = { .i32_value = val }
+                };
+            // fprintf(stderr, "using missing value %d => tag %c\n", val, value.tag);
```
Remove?

evanmiller commented on src/bin/modules/produce_csv_value_dta.c:
```diff
@@ -0,0 +1,106 @@
+                .tag = 'a' + i,
+                .v = { .double_value = val }
+                };
+            // fprintf(stderr, "using missing value %lf => tag %c\n", val, value.tag);
```
Remove line?

evanmiller commented on src/bin/readstat.c:
```diff
@@ -78,6 +48,9 @@ const char *format_name(int format) {
         return "SAS catalog file (SAS7BCAT)";
+    
+    if (format == RS_FORMAT_CSV)
+        return "CSV (NSD)";
```
What is NSD?

ivarref replied:
That would be the Norwegian abbreviation of Norwegian Centre for Research Data. I've changed it to simply say "CSV" now.

evanmiller commented on src/bin/write_value_labels.c:
```diff
@@ -0,0 +1,94 @@
+            char buf[255];
+            int k = value_label->int32_key;
+            char tag = 0;
+            if (k >= 2147483622) {
```
I think this number is the same as [DTA_113_MISSING_INT32_A](https://github.com/WizardMac/ReadStat/blob/master/src/stata/readstat_dta.h#L143), defined in [src/stata/readstat_dta.h](https://github.com/WizardMac/ReadStat/blob/master/src/stata/readstat_dta.h).

evanmiller commented on src/spss/readstat_sav_date.c:
```diff
@@ -0,0 +1,81 @@
+
+        days += day-1;
+        char buf[1024];
+        *dest = (char*)s + snprintf(buf, sizeof(buf)-1, "%d-%d-%d", year, month+1, day); 
```
Just FYI generally you'll want `snprintf(buf, sizeof(buf), ...)`, no need for `-1`. In this case it doesn't really matter though.

ivarref commented:
Thanks for the feedback @evanmiller . I've tried to correct what you mentioned.

I think my name should suffice for the patch ownership.

evanmiller commented:
Thanks @ivarref. The patch is looking good to me.

113 commits is a lot, would you mind doing a rebase before I merge? One massive commit is fine, though if you want a more careful curation that's okay too. [This page](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) has some useful info on rebasing.

ivarref commented:
Hi again @evanmiller 

It's been rebased (merge-squashed + reset --hard on master) now, and everything seems fine (`git diff master oldmaster` yields nothing).

Then I suppose it's all good to go? Thanks.

evanmiller commented:
Thanks @ivarref. Looks good to me, I'll merge and let other people have a go at it.

Thanks for sharing your work!

kylebarron commented:
Does this feature still exist in master? I'm trying to convert CSV to DTA files using readstat on the command line, but can't get a csv input accepted.

ivarref commented:
Hi @kylebarron 

I have not used the latest ReadStat, but the official docs says it still exists:
https://github.com/WizardMac/ReadStat#command-line-usage-with-csv-input

Basically you have to supply an additional metadata file when giving CSV as input.

Kind regards.

kylebarron commented:
I'd just missed the part about requiring `libcsv` to be found at compile time, and I would have expected an error to be raised at runtime, not just printing the help menu.
